src/transaction/atomic_replication_helper.cpp

```diff
@@ -268,35 +269,35 @@ namespace cublog
   atomic_replication_helper::atomic_log_sequence::add_atomic_replication_log (THREAD_ENTRY *thread_p,
       log_lsa record_lsa, LOG_RCVINDEX rcvindex, VPID vpid)
   {
-    m_log_vec.emplace_back (record_lsa, vpid, rcvindex);
-    const auto page_map_it = m_page_map.find (vpid);
-    if (page_map_it == m_page_map.cend ())
+    PAGE_PTR page_p = nullptr;
+    // bookkeeping fixes page, keeps all info regarding how the page was fixed (either
+    // regular fix or ordered fix) and only returns back the pointer to the page
+    const int err_code = m_page_ptr_bookkeeping.fix_page (thread_p, vpid, rcvindex, page_p);
+    if (err_code != NO_ERROR)
       {
-	int error_code = m_log_vec.back ().fix_page (thread_p);
-	if (error_code != NO_ERROR)
-	  {
-	    // failing to fix the page just leaves it unfixed and does not affect overall
-	    // functioning of the atomic replication sequence;
-	    // therefore, remove the log and move on
-	    m_log_vec.pop_back ();
-
-	    // TODO:
-	    //  - what happens if there is more than one log record pertaining to the same page
-	    //    in an atomic sequnce and, for example, for the first log record, the page fails
-	    //    to be fixed but succeeds for the second log record
-	    //  - what if, while the atomic sequence is in progress with a page having failed to
-	    //    be fixed, a client transactions manages to fix the page; IOW, how is the progress
-	    //    of the "highest processed LSA" working wrt atomic replication sequences
-
-	    return error_code;
-	  }
-	m_page_map.emplace (vpid,  m_log_vec.back ().get_page_ptr ());
+	// failing to fix the page just leaves it unfixed and does not affect overall
+	// functioning of the atomic replication sequence;
+	er_log_debug (ARG_FILE_LINE, "[ATOMIC_REPL] heap page cannot be fixed, cannot add new log record"
+		      "with LSA %lld|%d to atomic sequences started at LSA %lld|%d\n",
+		      LSA_AS_ARGS (&record_lsa), LSA_AS_ARGS (&m_start_lsa));
+
+	// TODO:
+	//  - what happens if there is more than one log record pertaining to the same page
+	//    in an atomic sequnce and, for example, for the first log record, the page fails
+	//    to be fixed but succeeds for the second log record
+	//  - what if, while the atomic sequence is in progress with a page having failed to
```

Comment:
For me, the safe approach would be: 
1. gather all log records of an atomic log seq
2. gather all VPID required
3. fix all pages
4. apply all log records
5. unfix all pages
I think the issue you mention may be caused by apply_and_unfix_sequence completing steps 4+5 at the same time and also if at step 3, if a fix fails we continue with the sequence. 
What if at least one page fix fails during step 3, we give up entirely and force evacuate all VPIDs from page buffer ?

Reply:
@razvanradulescu 

AFICT, this problem with the replication failing to fix a page while another client transaction might succeed is a possible scenario only in the context of page re-allocation. And there is an issue describing the problem and a possible solution which involves unilaterally aborting client transaction - see http://jira.cubrid.org/browse/LETS-523 .

Currently the atomic replication is implemented as follows:
1. when each log record deemed to be part of an atomic replication is processed, the page is fixed right then&there
  * the bookkeeping takes care whether two log records want to access the same page
2. when the section is concluded, all log records are applied and the page is unfixed *right* after each record is applied
  * again, bookkeeping takes care of when the ref count actually reaches 0 and the page *can* safely be unfixed by the overall sequence

The way I see there is not a fundamental difference between the way described in the previous comment and the implementation.
Except that the implementation does keep the pages fixed for longer. But that can be seen as a feature rather than a problem.
What if, while the replication keeps the page fixed (for a bit longer) an client transaction would need to fix the page for reading in order to read some data which - due to the MVCC functionality - would not collide with the actual data that the replication wants to modify.
Thus, one can argue, the current implementation might impact performance.